oracle/apioracle/apioracle.go

```diff
@@ -0,0 +1,241 @@
+package apioracle
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/vocdoni/storage-proofs-eth-go/ethstorageproof"
+	"github.com/vocdoni/storage-proofs-eth-go/token"
+	ethtoken "github.com/vocdoni/storage-proofs-eth-go/token"
+	"go.vocdoni.io/dvote/api"
+	"go.vocdoni.io/dvote/crypto/ethereum"
+	chain "go.vocdoni.io/dvote/ethereum"
+	ethereumhandler "go.vocdoni.io/dvote/ethereum/handler"
+	"go.vocdoni.io/dvote/log"
+	"go.vocdoni.io/dvote/oracle"
+	"go.vocdoni.io/dvote/router"
+	"go.vocdoni.io/dvote/util"
+	"go.vocdoni.io/dvote/vochain"
+	"go.vocdoni.io/proto/build/go/models"
+)
+
+const (
+	ethQueryTimeOut = 20 * time.Second
+)
+
+type APIoracle struct {
+	Namespace uint32
+	oracle    *oracle.Oracle
+	router    *router.Router
+	eh        *ethereumhandler.EthereumHandler
+}
+
+func NewAPIoracle(o *oracle.Oracle, r *router.Router) (*APIoracle, error) {
+	a := &APIoracle{router: r, oracle: o}
+	return a, nil
+}
+
+func (a *APIoracle) EnableERC20(chainName, web3Endpoint string) error {
+	if chainName == "" || web3Endpoint == "" {
+		return fmt.Errorf("no web3 endpoint or chain name provided")
+	}
+	specs, err := chain.SpecsFor(chainName)
+	if err != nil {
+		return err
+	}
+	a.eh, err = ethereumhandler.NewEthereumHandler(specs.Contracts, web3Endpoint)
+	if err != nil {
+		return err
+	}
+	a.eh.WaitSync()
+	a.router.RegisterPublic("newERC20process", a.handleNewEthProcess)
+	return nil
+}
+
+func (a *APIoracle) handleNewEthProcess(req router.RouterRequest) {
+	var response api.MetaResponse
+	if req.NewProcess == nil {
+		a.router.SendError(req, "newProcess is empty")
+		return
+	}
+	if req.NewProcess.EthIndexSlot == nil {
+		a.router.SendError(req, "index slot not provided")
+		return
+	}
+	if req.NewProcess.SourceHeight == nil {
+		a.router.SendError(req, "no source height provided")
+		return
+	}
+
+	pidseed := fmt.Sprintf("%d%d%x%x",
+		a.Namespace,
+		req.NewProcess.StartBlock,
+		req.NewProcess.EntityID,
+		util.RandomBytes(32),
+	)
+
+	p := &models.Process{
+		EntityId:          req.NewProcess.EntityID,
+		StartBlock:        req.NewProcess.StartBlock,
+		BlockCount:        req.NewProcess.BlockCount,
+		CensusRoot:        req.NewProcess.CensusRoot,
+		EnvelopeType:      req.NewProcess.EnvelopeType,
+		VoteOptions:       req.NewProcess.VoteOptions,
+		EthIndexSlot:      req.NewProcess.EthIndexSlot,
+		SourceBlockHeight: req.NewProcess.SourceHeight,
+		ProcessId:         ethereum.HashRaw([]byte(pidseed)),
+		Status:            models.ProcessStatus_READY,
+		Namespace:         a.Namespace,
+		CensusOrigin:      models.CensusOrigin_ERC20,
```

Comment:
to be future proof maybe we can check if is a supported token standard and not hardcode here the CensusOrigin_ERC20